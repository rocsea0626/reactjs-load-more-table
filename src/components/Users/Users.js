import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import moment from 'moment';
import {
  ErrorText,
  LoadingAnimation,
  LoadButton
} from '../../components';

const useStyles = makeStyles({
  table: {
    // minWidth: 650,
  },
  tableHeader: {
    fontWeight: 800
  },
  visuallyHidden: {
    border: 0,
    clip: 'rect(0 0 0 0)',
    height: 1,
    margin: -1,
    overflow: 'hidden',
    padding: 0,
    position: 'absolute',
    top: 20,
    width: 1,
  },
});

export const Users = (props) => {

  const classes = useStyles();

  const {
    order,
    orderBy,
    users,
    isLoading,
    error,
    handleOrderChanged,
    onLoadClicked
  } = props;


  const stableSort = (array, comparator) => {
    const stabilizedThis = array.map((el, index) => [el, index]);
    stabilizedThis.sort((a, b) => {
      const order = comparator(a[0], b[0]);
      if (order !== 0) return order;
      return a[1] - b[1];
    });
    return stabilizedThis.map((el) => el[0]);
  }

  const getComparator = (order, orderBy) => {
    return order === 'desc'
      ? (a, b) => descendingComparator(a, b, orderBy)
      : (a, b) => -descendingComparator(a, b, orderBy);
  }

  const descendingComparator = (a, b, orderBy) => {
    if (b[orderBy] < a[orderBy]) {
      return -1;
    }
    if (b[orderBy] > a[orderBy]) {
      return 1;
    }
    return 0;
  }

  return (
    <TableContainer component={Paper}>
      <Table
        data-testid="users-table"
        className={classes.table}
      >
        <TableHead >
          <TableRow>
            <TableCell className={classes.tableHeader}>
              <TableSortLabel
                data-testid="users-sort"
                active={true}
                direction={order}
                onClick={handleOrderChanged}
              >
                Date
                <span className={classes.visuallyHidden}>
                  {order === 'desc' ? 'sorted descending' : 'sorted ascending'}
                </span>
              </TableSortLabel>
            </TableCell>
            <TableCell className={classes.tableHeader}>User&nbsp;ID</TableCell>
            <TableCell className={classes.tableHeader}>Old&nbsp;value</TableCell>
            <TableCell className={classes.tableHeader}>New&nbsp;value</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {stableSort(users, getComparator(order, orderBy)).map((user) => (
            <TableRow
              key={user.id}
              data-testid="users-table-row"
            >
              <TableCell
                data-testid="users-table-row-date"
                component="th"
                scope="row"
              >
                {moment(user[orderBy]).format('YYYY-MM-DD')}
              </TableCell>
              <TableCell>{user.id}</TableCell>
              <TableCell>{user.diff[0].oldValue}</TableCell>
              <TableCell>{user.diff[0].newValue}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>

      <ErrorText
        data-testid="users-error-text"
        show={error ? true : false}
      />
      <LoadingAnimation
        data-testid="users-loading-anim"
        show={isLoading}
      />

      <LoadButton
        data-testid="users-load-btn"
        show={!isLoading}
        onClicked={onLoadClicked}
        hasError={error ? true : false}
      />

    </TableContainer>
  );
};

Users.propTypes = {
  orderBy: PropTypes.string.isRequired,
  order: PropTypes.string.isRequired,
  users: PropTypes.array.isRequired,
  isLoading: PropTypes.bool.isRequired,
  handleOrderChanged: PropTypes.func.isRequired,
  onLoadClicked: PropTypes.func.isRequired,
  error: PropTypes.object,
};

export default Users;