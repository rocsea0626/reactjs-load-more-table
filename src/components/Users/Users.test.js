import React from 'react';
import { shallow } from 'enzyme';
import { Users } from './Users';

describe('<Users />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Users />);
  });

  describe('render()', () => {
    it('renders the Box', () => {
      expect(wrapper.find({ 'data-testid': 'app-box' })).toHaveLength(1);
    });
  });
});
