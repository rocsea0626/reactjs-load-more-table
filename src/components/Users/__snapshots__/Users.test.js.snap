// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users /> snapshot test asc order 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root makeStyles-table-1"
    data-testid="users-table"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
            data-testid="users-sort"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Date
            <span
              className="makeStyles-visuallyHidden-3"
            >
              sorted descending
            </span>
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionDesc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          User ID
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          Old value
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          New value
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
    />
  </table>
  <div
    className="MuiBox-root MuiBox-root-149 Styled(MuiBox)-root-72"
  >
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      data-testid="loading-animation-box"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`<Users /> snapshot test desc order 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root makeStyles-table-1"
    data-testid="users-table"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
            data-testid="users-sort"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Date
            <span
              className="makeStyles-visuallyHidden-3"
            >
              sorted descending
            </span>
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionDesc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          User ID
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          Old value
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head makeStyles-tableHeader-2"
          scope="col"
        >
          New value
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
    />
  </table>
  <div
    className="MuiBox-root MuiBox-root-117 Styled(MuiBox)-root-72"
    data-testid="load-btn-box"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="load-btn"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Load more
      </span>
    </button>
  </div>
</div>
`;
